internal/dag/cache.go

```diff
@@ -84,24 +84,89 @@ func (kc *KubernetesCache) init() {
 	kc.extensions = make(map[types.NamespacedName]*contour_api_v1alpha1.ExtensionService)
 }
 
-// matchesIngressClass returns true if the given Kubernetes object
-// belongs to the Ingress class that this cache is using.
-func (kc *KubernetesCache) matchesIngressClass(obj metav1.Object) bool {
+// matchesIngressClass returns true if the given IngressClass
+// is the one this cache is using.
+func (kc *KubernetesCache) matchesIngressClass(obj *networking_v1.IngressClass) bool {
+	// If no ingress class name set, we allow an ingress class that is named
+	// with the default Contour accepted name.
+	if kc.IngressClassName == "" && obj.Name == "contour" {
+		return true
+	}
+	// Otherwise, the name of the ingress class must match what has been
+	// configured.
+	if obj.Name == kc.IngressClassName {
+		return true
+	}
+	return false
+}
+
+// ingressMatchesIngressClass returns true if the given Ingress object matches
+// the configured ingress class name via annotation or Spec.IngressClassName.
+// Annotation takes precedence over the Spec value.
+func (kc *KubernetesCache) ingressMatchesIngressClass(obj metav1.Object) bool {
+	annotationClass := annotation.IngressClass(obj)
+	var specClass string
+	switch obj := obj.(type) {
+	case *v1beta1.Ingress:
+		specClass = pointer.StringPtrDerefOr(obj.Spec.IngressClassName, "")
+	case *networking_v1.Ingress:
+		specClass = pointer.StringPtrDerefOr(obj.Spec.IngressClassName, "")
+	}
+
+	// If annotation is set or annotation is not set and spec is not set, we
+	// will try to match the annotation.
+	checkAnnotation := annotationClass != "" || specClass == ""
+
+	// If the annotation matches, return true.
+	if checkAnnotation && annotation.MatchesIngressClass(obj, kc.IngressClassName) {
```

Comment:
Yeah, we are not using the content of any `IngressClass` resource at all today. According to this it is an "optional" way to provide config which could be interpreted in a few ways (currently we have interpreted it such that a resource does not have to exist with the name referenced by an `Ingress`). We added read access to the resource but could totally walk it back if we want since we're not actually using it, though users might be annoyed if we have to eventually add RBAC to `IngressClass` back again at some point.

If we start making the `IngressClass` resource have to exist it may be an unexpected change for users who would have to go create before they upgrade or their `Ingress` routes will be dropped. We could require this now, if/when we actually look at the content of an `IngressClass`, or say we will never look at the content and never require it (maybe we could only require it you're using the spec field and not the annotation to specify the class? Seems like this might create more annoying edge cases and unexpected behavior, idk).

Reply:
I guess it wouldn't hurt to leave it as-is for now, and we can always add the existence requirement down the road, maybe once ingress v1beta1 is dropped?

Reply:
Even when it's only Ingress v1, the IngressClass is still optional, that's the API. I did not compute that until very recently either.